docs/2.10.0/docs/daml/troubleshooting.rst

```diff
@@ -33,6 +33,174 @@ Hovering over the compilation error displays:
 
   [Type checker] Argument expands to non-serializable type Party -> Party.
 
+Error: "Recursion limit overflow in module"
+*******************************************
+
+The error will usually occur when uploading a DAR to a ledger or using a Script
+via the sandbox. It can manifest on upload as
+
+.. code:: text
+
+  upload-dar did not succeed: DAR_PARSE_ERROR(8,b5935497): Failed to parse the dar file content.
+
+or in your logs as
+
+.. code:: text
+
+  Recursion limit overflow in module '<pkgid>:<modulename>'
+
+The cause of this error is usually having an expression in the DAR whose
+serialized representation exceeds a depth 1000 layers. This can be caused by
+long Daml scripts, since every use of a function call or ``<-`` to bind a
+variable in a ``do`` block can incur several layers of recursion in the ``do``
+block's serialized representation. This can also be caused by large datatypes
+with more than 160 fields with ``deriving`` clauses.
```

Comment:
```suggestion
This error is usually caused by having an expression in the DAR whose
serialized representation exceeds a depth of 1000 layers. This can be caused by
long Daml scripts, since every use of a function call or ``<-`` to bind a
variable in a ``do`` block can incur several layers of recursion in the ``do``
block's serialized representation. Large datatypes
with more than 160 fields with ``deriving`` clauses can also cause this.
```